### ISSUE TITLE
Tornado slower than expected with PyPy

### ISSUE THREAD

**Issue by gitlab-importer** [pypy/pypy#3402]
In Heptapod by @Adrian on Feb 19, 2021, 14:35

As you know, I've recently been troubleshooting a poorly-performing Tornado app, where PyPy was delivering 50% of CPython's throughput.

I managed to close this gap significantly (and drastically improve overall performance) by eliminating use of Tornado's "import" and "module" templating features.

I've created a minimal test case, where PyPy outperforms by 15%, but this is still far short of a [previously published benchmark](https://morepypy.blogspot.com/2017/03/async-http-benchmarks-on-pypy3.html) involving Tornado.

What's more, if I switch from Tornado's HTML output to JSON output, CPython outperforms by 50%, which I didn't expect.

I'm raising this issue in case you'd consider looking for any PyPy bottlenecks with one of the more popular frameworks. Minimal code below:

**app.py**

```
import tornado.httpserver
import tornado.ioloop
import tornado.web

from random import choice
from string import ascii_letters


text = [[''.join(choice(ascii_letters) for x in range(8)) for y in range(16)] for z in range(64)]


class MainHandler(tornado.web.RequestHandler):
    async def get(self):
        self.render("index.html", title="Test", text=text)
        # self.write(dict(title="Test", text=text))


def main():
    app = tornado.web.Application([(r"/", MainHandler)])
    app.listen(80)
    tornado.ioloop.IOLoop.current().start()


if __name__ == "__main__":
    main()
```

**index.html**

```
<!DOCTYPE html>
<html>
  <head>
    <title>{{ title }}</title>
  </head>
  <body>
    <h1>{{ title }}</h1>
    {% for par in text %}
      <p>{% for word in par %} {{ word }} {% end %}</p>
    {% end %}
  </body>
</html>
```

**Comment by mattip** [comment]
It would be nice to revive this work. As far as I can tell we did not merge the branches, which are (in git) branches/open-ended-traces and branches/open-ended-traces-3.7

**Comment by cfbolz** [comment]
open-ended-traces got merged in 5074f10752d9929e9d0570b3691804342692ff01, I think? seems we just forgot to close the issue, maybe? see also https://pypy.org/posts/2021/09/jit-auto-generated-code.html

**Comment by mattip** [comment]
Closing. Please open a new issue if there is more to do.
